[Form] Add ability to define form theme file from form type classes

bOmBeLq opened the pull request:
Enables ability to call:
```
public function configureOptions(OptionsResolver $resolver)
    {
      $resolver
          ->setDefault('theme', 'path/to/theme');
    }
```
on FormType class which will be equivalent to` {% form_theme form 'path/to/theme.{extension}'` where extension is `.html.twig` or `.html.php` depending on rendering engine.

Why?
1. Simplyfy definning views for specific form types.
2. Simplyfy defining form themes in case of re-usable bundles. Currently if your bundle has form types and views related to it you will have to use compiller pass to add your form theme to default themes list or leave note in docs saying to include the template somehow. By using form theme option this can be automated.

Original issue: #23070

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | will add if this feature gets accepted

Commit:
[Form] Add ability to define form theme file from form type classes

Enables ability to call:
public function configureOptions(OptionsResolver $resolver)
    {
      $resolver
          ->setDefault('theme', 'path/to/theme');
    }
which will be equivalent to {% form_theme form
'path/to/theme{extension}'
where extension is `.html.twig` or `.html.php` depending on rendering
engine.

Koc commented:
>  leave note in docs saying to include the template somehow

you can do it automatically in di-extension

bOmBeLq commented:
@Koc Yes as I wrote above you can use compiller pass. But I personally find this less convinient. 
Also you have to think of php and twig engines separatelly. With this patch you could define 
`/path/theme.html.twig`
`/path/theme.html.php`
and have proper one used basing on rendering engine. I didn't work on any other engines though so for any other template engine it would be ignored (easly possible to be added by registering the extension for any rendering engine).

You could also pretty easly change/set theme globally for any form type with `FormTypeExtension`. Though this is not my target with this change because you can define global form theme and use proper name for blocks to target desired form type.

netsuo commented:
I really like this

stof commented on CHANGELOG-3.3.md:
```diff
@@ -7,6 +7,9 @@ in 3.3 minor versions.
 To get the diff for a specific change, go to https://github.com/symfony/symfony/commit/XXX where XXX is the change hash
 To get the diff between two versions, go to https://github.com/symfony/symfony/compare/v3.3.0...v3.3.1
 
+* 3.3.7 (2017-08-18)
```
new features are never added in patch versions. and this file is autogenerated, so you should not edit it.

bOmBeLq replied:
Thank you @stof. I've reverted the changes on this file.

bOmBeLq commented on src/Symfony/Component/Form/Extension/Templating/Type/FormTypeThemeExtension.php:
```diff
@@ -0,0 +1,53 @@
+    {
+        $resolver
+            ->setDefault('theme', null)
+            ->setDefault('theme_extension', $this->extension);
```
@ogizanagi Why not sure? This pull request is actually all about this option. It would be fully backwards compatible anyway.

ogizanagi replied:
You already pass the extension in the constructor. I'm not sure about exposing the `theme_extension` option then, and I don't really think it should be changed.
BTW, what happen when you have both php and twig templating engines registered? This form type extension will be registered twice, right (one for each engine)? Depending on the order they are called, the `theme_extension` default will change unless I'm missing something.

bOmBeLq replied:
Oh you mean the file_extension. Sorry didn't get it.
I wanted to account the fact that someone may have some non-standard file extensions. If extension is left null then it is automatically defined basing on rendering engine (php/twig). But you can also force the extension yourself in case your template file has non-stanrad extension for any reason. 

Someone may use: `file_name.twig` format. In that case he can simply change the `theme_extension` to force `.twig` file extension.

bOmBeLq commented:
I've changed my mind on the `theme_extension` and removed it.
Instead I've added `theme_auto_extension` option.
By default it's value is `false` which means you would type full theme with exteions like so:
```
public function configureOptions(OptionsResolver $resolver)
    {
        $resolver
            ->setDefault('theme', 'path/to/theme.html.twig')
}
```
This will make it look more like path to the template. May also simplyfy refactoring. I bet this will cover like 99% cases.

Then with `theme_auto_extension` I've left doors open for people who would like to define templates for more then one templating engine. Setting `theme_auto_extension` will make it auto append default extension for current rendering engine (`.html.twig` or `.html.php`).
```
public function configureOptions(OptionsResolver $resolver)
    {
        $resolver
            ->setDefault('theme', 'path/to/theme') // notice no extension
           ->setDefault('theme_auto_extension', true);
}
```

nicolas-grekas commented:
Moving to 4.1 as we didn't manage to review/discussion on time.

bOmBeLq commented:
FYI for now I've created separate bundle for it if you would like to give it a try:
https://github.com/bOmBeLq/form-theme-bundle

xabbuh commented:
I am not convinced that having to define a form theme per form really is a common use case. I would vote to not merge this into the core. If someone needs this feature in the application, the bundle mentioned in https://github.com/symfony/symfony/pull/23990#issuecomment-362321411 seems to provide the proper solution. IMO that's enough.

ogizanagi commented:
Same opinion as @xabbuh

xabbuh commented:
I am going to close here then. Thank you for proposing this feature @bOmBeLq though it didn't make it this time.
